worldedit-core/src/main/java/com/boydti/fawe/beta/implementation/lighting/NMSRelighter.java

```diff
@@ -0,0 +1,637 @@
+package com.boydti.fawe.beta.implementation.lighting;
+
+import com.boydti.fawe.beta.IQueueChunk;
+import com.boydti.fawe.beta.IQueueExtent;
+import com.boydti.fawe.beta.implementation.chunk.ChunkHolder;
+import com.boydti.fawe.config.Settings;
+import com.boydti.fawe.object.RunnableVal;
+import com.boydti.fawe.object.collection.BlockVectorSet;
+import com.boydti.fawe.util.MathMan;
+import com.boydti.fawe.util.TaskManager;
+import com.sk89q.worldedit.math.MutableBlockVector3;
+import it.unimi.dsi.fastutil.longs.Long2ObjectOpenHashMap;
+
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Queue;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+public class NMSRelighter implements Relighter {
+    private final IQueueExtent<IQueueChunk> queue;
+
+    private final Map<Long, RelightSkyEntry> skyToRelight;
+    private final Object present = new Object();
+    private final Map<Long, Integer> chunksToSend;
+    private final ConcurrentLinkedQueue<RelightSkyEntry> extentdSkyToRelight = new ConcurrentLinkedQueue<>();
+
+    private final Map<Long, long[][][] /* z y x */ > lightQueue;
+    private final AtomicBoolean lightLock = new AtomicBoolean(false);
+    private final ConcurrentHashMap<Long, long[][][]> concurrentLightQueue;
+
+    private final int maxY;
+
+    public final MutableBlockVector3 mutableBlockPos = new MutableBlockVector3(0, 0, 0);
+
+    private static final int DISPATCH_SIZE = 64;
+    private boolean removeFirst;
+
+    public NMSRelighter(IQueueExtent<IQueueChunk> queue) {
+        this.queue = queue;
+        this.skyToRelight = new Long2ObjectOpenHashMap<>();
+        this.lightQueue = new Long2ObjectOpenHashMap<>();
+        this.chunksToSend = new Long2ObjectOpenHashMap<>();
+        this.concurrentLightQueue = new ConcurrentHashMap<>();
+        this.maxY = queue.getMaxY();
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return skyToRelight.isEmpty() && lightQueue.isEmpty() && extentdSkyToRelight.isEmpty() && concurrentLightQueue.isEmpty();
+    }
+
+    @Override
+    public synchronized void removeAndRelight(boolean sky) {
+        removeFirst = true;
+        fixLightingSafe(sky);
+        removeFirst = false;
+    }
+
+    private void set(int x, int y, int z, long[][][] map) {
```

Comment:
A vague method with the name "set" should have a comment before it explaining what the method does because it isn't clear what is being set.